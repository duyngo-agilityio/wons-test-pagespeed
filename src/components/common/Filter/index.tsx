'use client';

import { memo } from 'react';
import isEqual from 'react-fast-compare';
import {
  Button,
  Popover,
  PopoverTrigger,
  PopoverContent,
  Listbox,
  ListboxItem,
  ButtonProps,
  PopoverProps,
  PopoverTriggerProps,
  ListboxProps,
  ListboxItemProps,
  PopoverContentProps,
} from '@nextui-org/react';

// Types
import { IFilter } from '@/types';

// Icons
import { FilterIcon } from '../icons';

// Components
import { Checkbox, Text } from '@/components/common';

interface IFilterProps {
  items: IFilter[];
  title?: string;
  buttonProps?: ButtonProps;
  popoverProps?: PopoverProps;
  popoverTriggerProps?: PopoverTriggerProps;
  popoverContentProps?: PopoverContentProps;
  listboxProps?: ListboxProps;
  listboxItemProps?: ListboxItemProps;
}

const Filter = ({
  items,
  title = '',
  buttonProps,
  popoverProps,
  popoverTriggerProps,
  popoverContentProps,
  listboxProps,
  listboxItemProps,
}: IFilterProps) => {
  return (
    <Popover {...popoverProps}>
      <PopoverTrigger {...popoverTriggerProps}>
        {title && (
          <Button
            className="bg-blue-500 border-transparent dark:bg-purple-600 text-white hover:bg-blue-100 px-5 py-3 text-md rounded-[10px] border-[1px] h-auto min-w-max font-medium"
            endContent={<FilterIcon />}
            {...buttonProps}
          >
            {title}
          </Button>
        )}
      </PopoverTrigger>
      <PopoverContent className="p-3" {...popoverContentProps}>
        {items.map(({ id, title, items }) => (
          <div className="w-full" key={id}>
            <Text
              text={title}
              className="font-medium border-b-1 border-blue-800/50 pb-1"
            />
            <Listbox {...listboxProps}>
              {items.map(({ id, content, customElement }) => (
                <ListboxItem key={id} className="px-0" {...listboxItemProps}>
                  {customElement || <Checkbox>{content}</Checkbox>}
                </ListboxItem>
              ))}
            </Listbox>
          </div>
        ))}
      </PopoverContent>
    </Popover>
  );
};

export default memo(Filter, isEqual);
