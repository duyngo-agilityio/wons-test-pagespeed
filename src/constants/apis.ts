export const API_BASE_URL = process.env.NEXT_API_BASE_URL || '';

export const API_PATH = {
  SIGN_UP: '/auth/local/register',
  USERS: '/users',
};
