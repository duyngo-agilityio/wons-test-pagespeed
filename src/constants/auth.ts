// Types
import { TRole } from '@/types';

export const AUTH_METHODS = {
  CREDENTIALS: 'credentials',
};

export const ERROR_TYPES = {
  CREDENTIALS_SIGN_IN: 'CallbackRouteError',
};

export const ROLES: TRole[] = [
  {
    id: 3,
    name: 'Admin',
  },
  {
    id: 4,
    name: 'User',
  },
];
