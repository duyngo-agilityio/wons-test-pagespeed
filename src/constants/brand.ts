export const BRANDS = [
  { key: 'apple', label: 'Apple' },
  { key: 'samsung', label: 'Samsung' },
  { key: 'huawei', label: 'Huawei' },
  { key: 'xioami', label: 'Xioami' },
  { key: 'oppo', label: 'Oppo' },
  { key: 'google', label: 'Google' },
];
