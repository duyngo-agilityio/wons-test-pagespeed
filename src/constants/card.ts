export enum LevelCard {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
}
