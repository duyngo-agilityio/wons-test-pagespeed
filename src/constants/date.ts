export const DEFAULT_RANGE_VALUE_PICKER = 7;
