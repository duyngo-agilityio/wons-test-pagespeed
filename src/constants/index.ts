export * from './apis';
export * from './auth';
export * from './date';
export * from './fonts';
export * from './images';
export * from './messages';
export * from './params';
export * from './regex';
export * from './routes';
export * from './sidebar';
export * from './themes';
export * from './typography';
export * from './messages';
export * from './regex';
export * from './apis';
export * from './auth';
export * from './invoices';
export * from './select';
export * from './pagination';
