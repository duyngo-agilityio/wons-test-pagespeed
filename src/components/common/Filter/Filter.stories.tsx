// Libs
import type { Meta, StoryObj } from '@storybook/react';

// Constants
import { FILTER_OPTIONS } from '@/constants';

// Components
import Filter from './index';

const meta: Meta<typeof Filter> = {
  title: 'Components/Common/Filter',
  component: Filter,
  args: {
    items: FILTER_OPTIONS,
  },
  argTypes: {
    items: {
      description:
        'An array of filter items, each containing an id, title, and an array of items.',
    },
    title: {
      type: 'string',
      description: 'The title of the filter button.',
    },
    buttonProps: {
      description: 'Additional props to pass to the filter button.',
    },
    popoverProps: {
      description: 'Additional props to pass to the popover container.',
    },
    popoverTriggerProps: {
      description: 'Additional props to pass to the popover trigger.',
    },
    popoverContentProps: {
      description: 'Additional props to pass to the popover content.',
    },
    listboxProps: {
      description: 'Additional props to pass to the listbox container.',
    },
    listboxItemProps: {
      description: 'Additional props to pass to each listbox item.',
    },
  },
};

export default meta;
type Story = StoryObj<typeof Filter>;

export const Default: Story = {
  args: {
    title: 'Filter',
  },
};
