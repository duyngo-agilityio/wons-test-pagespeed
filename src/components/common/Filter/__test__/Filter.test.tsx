// Constants
import { FILTER_OPTIONS } from '@/constants';

// Components
import Filter from '../index';

describe('Filter component', () => {
  const renderComponent = () =>
    testLibJestUtils.render(<Filter title="Filter" items={FILTER_OPTIONS} />);

  it('should match snapshot', () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });
});
