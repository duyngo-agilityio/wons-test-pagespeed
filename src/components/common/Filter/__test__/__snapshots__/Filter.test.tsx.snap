// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component should match snapshot 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
    <p
      aria-expanded="false"
      class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-slot="trigger"
      type="button"
    />
  </div>
</div>
`;
